File: src/j/tests/MicroSet_Benchmark_Add.java
package j.tests;


import com.google.caliper.BeforeExperiment;
import com.google.caliper.Benchmark;
import com.google.caliper.api.VmOptions;
import com.google.caliper.runner.CaliperMain;
import j.generator.Generator;
import j.implementation.TestObject;
import j.implementation.microSet.AbstractInnerSet;
import j.implementation.microSet.MicroSet;
import j.implementation.microSet.InnerArraySet;

import java.util.HashSet;
import java.util.Random;
import java.util.TreeSet;

@VmOptions("-XX:-TieredCompilation")
public class MicroSet_Benchmark_Add {

    public static final int NUMBER_OF_TEST_OBJECT = 10;

    MicroSet<TestObject> microInnerSet = new MicroSet<>();
    MicroSet<TestObject> microArraySet = new MicroSet<>();
    MicroSet<TestObject> microHashSet = new MicroSet<>();
    HashSet<TestObject> hashSet = new HashSet<>();
    TreeSet<TestObject> treeSet = new TreeSet<>();

    TestObject[] testObjects = new TestObject[NUMBER_OF_TEST_OBJECT];
    int[] randomInt = new int[NUMBER_OF_TEST_OBJECT];

    @BeforeExperiment
    void setUp() {
        Random random = new Random();
        for (int i = 0; i < NUMBER_OF_TEST_OBJECT; i++) {
            testObjects[i] = new TestObject(random.nextInt());
        }
        for (int i = 0; i < NUMBER_OF_TEST_OBJECT; i++) {
            randomInt[i] = random.nextInt(NUMBER_OF_TEST_OBJECT);
        }
    }


    @Benchmark
    public void testAdd_TreeSet(int reps) {
        for (int i = 0; i < reps; i++) {
            for (int j = 0; j < NUMBER_OF_TEST_OBJECT; j++) {
                treeSet.add(testObjects[randomInt[j]]);
            }
        }
    }

    @Benchmark
    public void testAdd_HashSet(int reps) {
        for (int i = 0; i < reps; i++) {
            for (int j = 0; j < NUMBER_OF_TEST_OBJECT; j++) {
                hashSet.add(testObjects[randomInt[j]]);
            }
        }
    }

    @Benchmark
    public void testAdd_MicroHashSet(int reps) {
        MicroSet.use = MicroSet.Use.HASH_SET;
        microHashSet = new MicroSet<>();
        for (int i = 0; i < reps; i++) {
            for (int j = 0; j < NUMBER_OF_TEST_OBJECT; j++) {
                microHashSet.add(testObjects[randomInt[j]]);
            }
        }
    }

    @Benchmark
    public void testAdd_MicroArraySet(int reps) {
        MicroSet.use = MicroSet.Use.ARRAY_SET;
        microArraySet = new MicroSet<>();
        for (int i = 0; i < reps; i++) {
            for (int j = 0; j < NUMBER_OF_TEST_OBJECT; j++) {
                microArraySet.add(testObjects[randomInt[j]]);
            }
        }
    }
    @Benchmark
    public void testAdd_MicroInnerSet(int reps) {
        MicroSet.use = MicroSet.Use.INNER_SET;
        microInnerSet = new MicroSet<>();
        for (int i = 0; i < reps; i++) {
            for (int j = 0; j < NUMBER_OF_TEST_OBJECT; j++) {
                microInnerSet.add(testObjects[randomInt[j]]);
            }
        }
    }

    public static void main(String[] args) {
        args = new String[]{ "-i", "runtime", "-r", "ADD scale=" + Generator.TO_GENERATE+ ", object_number=" + NUMBER_OF_TEST_OBJECT};
        CaliperMain.main(MicroSet_Benchmark_Add.class, args);
    }
}

File: src/scala/tests/MicroSet_Benchmark_Add.scala
package scala.tests

import java.util.{HashSet, Random, TreeSet}

import com.google.caliper.{BeforeExperiment, Benchmark}
import com.google.caliper.api.VmOptions
import com.google.caliper.runner.CaliperMain

import scala.implementation.{MicroSet, Use}

/**
  * Created by pxl on 03/04/17.
  */

@VmOptions (Array ("-XX:-TieredCompilation") )
class MicroSet_Benchmark_Add {

  val microInnerSet = new MicroSet[TestObject]()
  MicroSet.use = Use.ARRAY_SET
  val microArraySet = new MicroSet[TestObject]()
  MicroSet.use = Use.HASH_SET
  val microHAshSet = new MicroSet[TestObject]()

  val hashSet = new HashSet[TestObject]
  val treeSet = new TreeSet[TestObject]

  var tabTestObject: Array[TestObject] = Array.ofDim(runner.NUMBER_OF_TEST_OBJECT)
  var tabRandomInt: Array[Int] = Array.ofDim(runner.NUMBER_OF_TEST_OBJECT)

  @BeforeExperiment()
  def setUp() =
  {
    val rand: Random = new Random
    for(i <- 0 until runner.NUMBER_OF_TEST_OBJECT) tabTestObject(i) = new TestObject(rand.nextInt(13))
    for(i <- 0 until runner.NUMBER_OF_TEST_OBJECT) tabRandomInt(i) = rand.nextInt(runner.NUMBER_OF_TEST_OBJECT)
  }

  @Benchmark
  def testAdd_TreeSet(reps: Int) =
  {
     for(i <- 0 to reps)
     {
       for(j <- 0 until runner.NUMBER_OF_TEST_OBJECT) treeSet.add(tabTestObject(tabRandomInt(j)))
     }
  }

  @Benchmark
  def testAdd_HashSet(reps: Int) =
  {
    for(i <- 0 to reps)
    {
      for(j <- 0 until runner.NUMBER_OF_TEST_OBJECT) hashSet.add(tabTestObject(tabRandomInt(j)))
    }
  }

  @Benchmark
  def testAdd_InnerSet(reps: Int) =
  {
    for(i <- 0 to reps)
    {
      for(j <- 0 until runner.NUMBER_OF_TEST_OBJECT) microInnerSet.add(tabTestObject(tabRandomInt(j)))
    }
  }

  @Benchmark
  def testAdd_InnerArraySet(reps: Int) =
  {
    for(i <- 0 to reps)
    {
      for(j <- 0 until runner.NUMBER_OF_TEST_OBJECT) microArraySet.add(tabTestObject(tabRandomInt(j)))
    }
  }

  @Benchmark
  def testAdd_InnerHashSet(reps: Int) =
  {
    for(i <- 0 to reps)
    {
      for(j <- 0 until runner.NUMBER_OF_TEST_OBJECT) microHAshSet.add(tabTestObject(tabRandomInt(j)))
    }
  }
}

object runner {
  val NUMBER_OF_TEST_OBJECT: Int = 30
  def main(args: Array[String]): Unit = {
    val args = Array[String]("-i", "runtime", "-r", "ADD scale=" + NUMBER_OF_TEST_OBJECT + ", object_number=" + NUMBER_OF_TEST_OBJECT)
    CaliperMain.main(classOf[MicroSet_Benchmark_Add], args)
  }
}
